libnettest2.hpp

```diff
@@ -0,0 +1,1805 @@
+// the interface and implementation using LIBNETTEST2_NO_INLINE_IMPL.
+#ifndef LIBNETTEST2_NO_INLINE_IMPL
+
+// Errors
+// ``````
+
+const char *libnettest2_strerror(Errors n) noexcept {
+#define XX(e_) case Errors::e_: return #e_;
+  switch (n) {
+    LIBNETTEST2_ENUM_OWN_ERRORS(XX)
+  }
+#undef XX
+  return "invalid_argument";
+}
+
+// UUID4 code
+// ``````````
+// Derivative work of r-lyeh/sole@c61c49f10d.
+/*-
+ * Portions Copyright (c) 2015 r-lyeh (https://github.com/r-lyeh)
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ * claim that you wrote the original software. If you use this software
+ * in a product, an acknowledgment in the product documentation would be
+ * appreciated but is not required.
+ *
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ * misrepresented as being the original software.
+ *
+ * 3. This notice may not be removed or altered from any source distribution.
+ */
+namespace sole {
+
+class uuid {
+  public:
+    std::string str();
+    uint64_t ab;
+    uint64_t cd;
+};
+
+uuid uuid4();
+
+std::string uuid::str() {
+  std::stringstream ss;
+  ss << std::hex << std::nouppercase << std::setfill('0');
+
+  uint32_t a = (ab >> 32);
+  uint32_t b = (ab & 0xFFFFFFFF);
+  uint32_t c = (cd >> 32);
+  uint32_t d = (cd & 0xFFFFFFFF);
+
+  ss << std::setw(8) << (a) << '-';
+  ss << std::setw(4) << (b >> 16) << '-';
+  ss << std::setw(4) << (b & 0xFFFF) << '-';
+  ss << std::setw(4) << (c >> 16) << '-';
+  ss << std::setw(4) << (c & 0xFFFF);
+  ss << std::setw(8) << d;
+
+  return ss.str();
+}
+
+uuid uuid4() {
+  std::random_device rd;
+  std::uniform_int_distribution<uint64_t> dist(0, (uint64_t)(~0));
+  uuid my;
+
+  my.ab = dist(rd);
+  my.cd = dist(rd);
+
+  /* The version 4 UUID is meant for generating UUIDs from truly-random or
+     pseudo-random numbers.
+
+     The algorithm is as follows:
+
+     o  Set the four most significant bits (bits 12 through 15) of the
+        time_hi_and_version field to the 4-bit version number from
+        Section 4.1.3.
+
+     o  Set the two most significant bits (bits 6 and 7) of the
+        clock_seq_hi_and_reserved to zero and one, respectively.
+
+     o  Set all the other bits to randomly (or pseudo-randomly) chosen
+        values.
+
+     See <https://tools.ietf.org/html/rfc4122#section-4.4>. */
+  my.ab = (my.ab & 0xFFFFFFFFFFFF0FFFULL) | 0x0000000000004000ULL;
+  my.cd = (my.cd & 0x3FFFFFFFFFFFFFFFULL) | 0x8000000000000000ULL;
+
+  return my;
+}
+
+}  // namespace sole
```

Comment:
Shouldn't this maybe be vendored via a separate header?

Note: we are soon going to be dropping client-side generated id in favour of server-side measurement ID

Reply:
> Shouldn't this maybe be vendored via a separate header?

I'd rather not do that, for two reasons: (1) this is code that I modified substantially; (2) I like to have a single header unless circumstances force me to separate it into smaller headers, because that simplifies integration.

The latter problem is solved by `nlohmann/json` and others using `amalgamate` to glue smaller headers together but I'd argue that currently this library header is quite manageable in size.

> Note: we are soon going to be dropping client-side generated id in favour of server-side measurement ID

Good, when that will be true, I'll zap this chunk of code from libnettest2.

Reply:
Added clarification in b3b966d1575b1050e044c8d4c06882b3105c9bfb